cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -410,4 +410,45 @@ const char* convert_fromUTF8_binary(int64 context, const char* bin_in, int32 len
   return ret;
 }
 
+FORCE_INLINE
+const char* replace_utf8_utf8_utf8(int64 context, const char* text, int32 text_len,
+                                   const char* from_str, int32 from_str_len,
+                                   const char* to_str, int32 to_str_len, int32* out_len) {
+  // if from_str is empty or its length exceeds that of original string,
+  // return the original string
+  if (from_str_len == 0 || from_str_len > text_len) {
+    *out_len = text_len;
+    return text;
+  }
+
+  char* out = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, 8000));
+  if (out == nullptr) {
+    gdv_fn_context_set_error_msg(context, "Could not allocate memory for output string");
+    *out_len = 0;
+    return "";
+  }
+
+  int32 out_index = 0;
+  int32 i = 0;
+  for (; i <= text_len - from_str_len;) {
+    int32 j = 0;
+    for (; j < from_str_len; ++j) {
```

Comment:
especially if this is going to be used repeatedly, I would expect you might be able to get performance gains by implementing https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm (assuming you can store the analyzed word in context)

Reply:
Or maybe Boyer-Moore, or another algorithm from https://en.wikipedia.org/wiki/String-searching_algorithm

Reply:
I can try implementing KMP or BM. There's also the Boyer Moore search algorithm in boost, (would need to convert the char array to std::string which would do a copy) . Also there's c library function strstr which does pretty well[1] considering it's heavily
optimised (but it expects null terminated strings so I have to do a memcpy). Any thoughts?

[1] https://stackoverflow.com/questions/7586990/strstr-faster-than-algorithms

Reply:
I haven't done research into existing implementations, but I think requiring a memcpy would lose a lot of the benefit of this algorithm. At least half on bm is fairly easy to implement.  Maybe start there and open up a follow-up JIRA.

Reply:
Ok, I will create a followup ticket for implementing BM. Meanwhile we can land this PR. I have added commit addressing the remaining comments.